### PR TITLE
[RFC] Fix potential deadlock

### PR DESCRIPTION
**Pull request by erw7**
`ELOG` may call `os_getenv()` and `os_setenv()` internally. In that case, a deadlock occurs.

### CODE REVIEW

File: src/nvim/os/env.c [hunk 1]
```diff
@@ -74,16 +74,16 @@ const char *os_getenv(const char *name)
   }
   if (r != 0 || size == 0 || e[0] == '\0') {
     xfree(e);
-    e = NULL;
+    uv_mutex_unlock(&mutex);
```
Comment (justinmk) [inline src/nvim/os/env.c]: // Must do this before ELOG, log.c may call os_setenv.

**Comment by justinmk** [comment]
In all cases it looks like ELOG  could just be moved to the end, instead of repeating the unlock code .

File: src/nvim/os/env.c [hunk 2]
```diff
@@ -67,22 +68,23 @@ const char *os_getenv(const char *name)
...
     goto end;
   }
   pmap_put(cstr_t)(envmap, xstrdup(name), e);
 end:
+  // Must do this before ELOG, log.c may call os_setenv.
   uv_mutex_unlock(&mutex);
+  if (r != 0 && r != UV_ENOENT && r != UV_UNKNOWN) {
+    ELOG("uv_os_getenv(%s) failed: %d %s", name, r, uv_err_name(r));
+  }
   return (e == NULL || size == 0 || e[0] == '\0') ? NULL : e;
```
Comment (erw7) [inline src/nvim/os/env.c]: @justinmk Is this condition necessary? I think good with `return e;`.

Comment (justinmk) [reply src/nvim/os/env.c]: Would need to check the vcs logs. Not really worth the trouble to change it...

**Review by justinmk — APPROVED** [review]

**Comment by blueyed** [comment]
LGTM.  Do you have a way to trigger/test it easily?  How did you notice?

**Comment by erw7** [comment]
> Do you have a way to trigger/test it easily?

To test this, we need to use a libuv that implements a special `uv_os_getenv()`, `uv_os_setenv()` that returns an error during the test. Therefore, there is no easy way to test this.

> How did you notice?

I didn't comment because it was a WIP, but I noticed the possibility because the https://github.com/neovim/neovim/pull/11140/commits/bdde40b30290cb38388c0eab79d587f0b324a462 test was stalled.

**Pull request merged**